components/brave_extension/extension/brave_extension/background/greaselion.ts

```diff
@@ -372,49 +438,77 @@ chrome.runtime.onConnectExternal.addListener((port: chrome.runtime.Port) => {
                   error
                 }
               })
-            })
-          break
-        }
-        case 'RegisterOnCompletedWebRequest': {
-          const data = msg.data as RegisterOnCompletedWebRequest
-          handleRegisterOnCompletedWebRequest(key, msg.mediaType, data)
-          break
-        }
-        case 'RegisterOnSendHeadersWebRequest': {
-          const data = msg.data as RegisterOnSendHeadersWebRequest
-          handleRegisterOnSendHeadersWebRequest(key, msg.mediaType, data)
-          break
-        }
-        case 'RegisterOnUpdatedTab': {
-          handleRegisterOnUpdatedTab(key, msg.mediaType)
-          break
-        }
-        case 'SavePublisherVisit': {
-          const data = msg.data as SavePublisherVisit
-          handleSavePublisherVisit(tabId, msg.mediaType, data)
-          break
-        }
-        case 'TipUser': {
-          const data = msg.data as TipUser
-          handleTipUser(tabId, msg.mediaType, data)
-          break
-        }
+            }
+          })
+        break
       }
-    })
+      case 'RegisterOnCompletedWebRequest': {
+        const data = msg.data as RegisterOnCompletedWebRequest
+        handleRegisterOnCompletedWebRequest(key, msg.mediaType, data)
+        break
+      }
+      case 'RegisterOnSendHeadersWebRequest': {
+        const data = msg.data as RegisterOnSendHeadersWebRequest
+        handleRegisterOnSendHeadersWebRequest(key, msg.mediaType, data)
+        break
+      }
+      case 'RegisterOnUpdatedTab': {
+        handleRegisterOnUpdatedTab(key, msg.mediaType)
+        break
+      }
+      case 'SavePublisherVisit': {
+        const data = msg.data as SavePublisherVisit
+        handleSavePublisherVisit(tabId, msg.mediaType, data)
+        break
+      }
+      case 'TipUser': {
+        const data = msg.data as TipUser
+        handleTipUser(tabId, msg.mediaType, data)
+        break
+      }
+    }
   })
+}
+
+chrome.runtime.onConnectExternal.addListener((port: chrome.runtime.Port) => {
+  if (!port || port.name !== 'Greaselion') {
+    return
+  }
+
+  port.onMessage.addListener(onMessageListener)
 
   port.onDisconnect.addListener((port: chrome.runtime.Port) => {
     if (chrome.runtime.lastError) {
       console.error(`Greaselion port disconnected due to error: ${chrome.runtime.lastError}`)
     }
     if (port.sender && port.sender.id && port.sender.tab && port.sender.tab.id) {
+      const key = buildTabIdSenderIdKey(port.sender.tab.id, port.sender.id)
+
+      const connectionState = connectionsByTabIdSenderId.get(key)
+      if (connectionState) {
+        onCompletedWebRequestRegistrations.delete(key)
```

Comment:
If you *wanted* to, you could probably remove these maps, since I think we can now get the same information from the connection state object. 

```js
if (onCompletedWebRequestRegistrations.has(key) && 
    onCompletedWebRequestRegistrations.get(key).registered) {
  assert(connectionState.onCompletedWebRequestListener);
}
```

Reply:
Great idea, done.